power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -714,52 +766,174 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
+                        bool all_match = true;
+                        for (Idx i = 0; i < elements_ps; ++i) {
+                            for (Idx j = 0; j < n_scenario_comp_in_update; ++j) {
+                                if (id_buffer[i] != id_buffer[i + elements_ps * j]) {
+                                    all_match = false;
+                                    break;
+                                }
+                            }
+                            if (!all_match) {
+                                break;
+                            }
+                        }
+                        return all_match;
+                    }();
+                    if (comp_info.elements_per_scenario != invalid_index) {
+                        result.elements_ps_in_update = comp_info.elements_per_scenario;
+                    }
+                } else {                        // no id,
+                    result.ids_all_na = true;   // no id, all NA
+                    result.ids_part_na = false; // no id, no part NA
+                    result.ids_match = true;    // no id, all match
+                }
+            } else {
+                process_buffer_span.template operator()<CT>(
+                    update_data.get_buffer_span_all_scenarios<meta_data::update_getter_s, CT>(), result);
+            }
+            if (comp_index >= 0 && result.uniform) {
+                result.elements_ps_in_update =
+                    update_data.get_component_info(comp_index).elements_per_scenario; // -1 for sparse
+            }
+            return result;
         };
 
-        // check all components
-        auto const update_independent = run_functor_with_all_types_return_array(is_component_update_independent);
-        return std::ranges::all_of(update_independent, [](bool const is_independent) { return is_independent; });
+        // check and return indenpendence of all components
+        return run_functor_with_all_types_return_vector(check_each_component);
+    }
+
+    std::vector<std::pair<std::string, bool>> is_update_independent(ConstDataset const& update_data) {
+        std::vector<std::pair<std::string, bool>> result;
+
+        // If the batch size is (0 or) 1, then the update data for this component is 'independent'
+        if (update_data.batch_size() <= 1) {
+            result.emplace_back("all component", true);
+            return result;
+        }
+
+        auto const all_comp_update_independence = check_components_independence(update_data);
+        for (auto const& comp : all_comp_update_independence) {
+            result.emplace_back(comp.name, comp.is_independent());
+        }
+
+        return result;
+    }
+    void validate_update_data_independence(UpdateCompProperties const& comp) const {
+        if (!comp.has_id && comp.ids_all_na) {
+            return; // empty dataset is still supported
```

Comment:
i think this can also be a `constexpr` function of the `UpdateCompProperties`

Reply:
For refactoring.